bin/propolis-server/src/lib/spec/api_spec_v0.rs

```diff
@@ -0,0 +1,273 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Conversions from version-0 instance specs in the [`propolis_api_types`]
+//! crate to the internal [`super::Spec`] representation.
+
+use std::collections::HashMap;
+
+use propolis_api_types::instance_spec::{
+    components::devices::{SerialPort as SerialPortDesc, SerialPortNumber},
+    v0::{InstanceSpecV0, NetworkBackendV0, NetworkDeviceV0, StorageDeviceV0},
+};
+use thiserror::Error;
+
+#[cfg(feature = "falcon")]
+use propolis_api_types::instance_spec::components::devices::SoftNpuPort as SoftNpuPortSpec;
+
+#[cfg(feature = "falcon")]
+use crate::spec::SoftNpuPort;
+
+use super::{
+    builder::{SpecBuilder, SpecBuilderError},
+    Disk, Nic, QemuPvpanic, SerialPortDevice, Spec,
+};
+
+#[derive(Debug, Error)]
+pub(crate) enum ApiSpecError {
+    #[error(transparent)]
+    Builder(#[from] SpecBuilderError),
+
+    #[error("backend {backend} not found for device {device}")]
+    BackendNotFound { backend: String, device: String },
+
+    #[error("backend {0} not used by any device")]
+    BackendNotUsed(String),
+
+    #[error("network backend for guest NIC {0} is not a viona backend")]
+    GuestNicInvalidBackend(String),
+
+    #[cfg(feature = "falcon")]
+    #[error("network backend for device {0} is not a DLPI backend")]
+    NotDlpiBackend(String),
+}
+
+impl From<Spec> for InstanceSpecV0 {
+    fn from(val: Spec) -> Self {
+        // Inserts a component entry into the supplied map, asserting first that
+        // the supplied key is not present in that map.
+        //
+        // This assertion is valid because internal instance specs should assign
+        // a unique name to each component they describe. The spec builder
+        // upholds this invariant at spec creation time.
+        fn insert_component<T>(
```

Comment:
nit: maybe worth sticking a 
```suggestion
        #[track_caller]
        fn insert_component<T>(
```

Reply:
whoaaaaa